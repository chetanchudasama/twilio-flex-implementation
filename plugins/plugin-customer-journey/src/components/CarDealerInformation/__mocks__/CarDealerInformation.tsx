import React from "react";
import { Props } from "../CarDealerInformation";

const CarDealerInformation: React.FC<Props> = (props) => <div />;

export default CarDealerInformation;
