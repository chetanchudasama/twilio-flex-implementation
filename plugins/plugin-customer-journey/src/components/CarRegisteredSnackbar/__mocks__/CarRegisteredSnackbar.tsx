import React from "react";
import { CarRegisteredSnackbarProps } from "../CarRegisteredSnackbar.Props";

export const CarRegisteredSnackbar: React.FC<CarRegisteredSnackbarProps> = (
  props
) => {
  return <div className="car-registered-snackbar">CarRegisteredSnackbar</div>;
};
