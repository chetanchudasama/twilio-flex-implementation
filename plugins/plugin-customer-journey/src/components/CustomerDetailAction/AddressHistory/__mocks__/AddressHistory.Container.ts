import AddressHistory from "./AddressHistory";

export default AddressHistory;
