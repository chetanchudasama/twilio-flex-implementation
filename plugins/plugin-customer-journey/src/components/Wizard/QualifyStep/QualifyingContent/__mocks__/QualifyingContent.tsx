import React from "react";
import { QualifyingContentProps } from "../QualifyingContent";

const QualifyingContent: React.FC<QualifyingContentProps> = (_props) => <div />;

export default QualifyingContent;
