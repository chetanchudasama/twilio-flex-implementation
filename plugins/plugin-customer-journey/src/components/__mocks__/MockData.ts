import { WorkerAttributes } from "@twilio/flex-ui";
import { namespace } from "../../states";

const workerAttributes: WorkerAttributes = {
  contact_uri: "client:test",
  email: "test@test.com",
  full_name: "test user",
  image_url: "img_url",
  roles: ["admin"],
  agentId: 1,
};

const conferenceSource = {
  source: {
    participants: [
      {
        callSid: "callSid",
        participantType: "customer",
        workerSid: "",
      },
      {
        callSid: "callSid",
        participantType: "worker",
        workerSid: "workerSid",
      },
    ],
  },
};

const conferenceState = new Map();
conferenceState.set("taskSid", conferenceSource);

export const flexManagerMock = jest.fn().mockReturnValue({
  workerClient: {
    source: {},
    sid: "workerSid",
    name: "test user",
    fullName: "test user",
    activityName: "Offline",
    attributes: workerAttributes,
    dateUpdated: new Date(),
    isAvailable: false,
    activityDuration: "",
    dateStatusChanged: new Date(1).toISOString(),
    activity: {
      available: false,
      name: "Offline",
      sid: "OfflineSid",
    },
  },
  store: {
    getState: jest.fn().mockReturnValue({
      flex: {
        worker: {
          attributes: workerAttributes,
        },
        session: {
          ssoTokenPayload: {
            token: "ssoToken",
          },
        },
        conferences: {
          states: conferenceState,
        },
      },
      [namespace]: {
        phoneNumberList: {
          phoneNumberList: [
            {
              friendlyName: "Sales",
              phoneNumber: "+1234567891",
              queueSid: "QUEUE_SID",
            },
          ],
        },
      },
    }),
  },
  strings: {
    DIALPADExternalTransferSearchInputLabel: "Search By Name or PhoneNumber",
    DIALPADExternalTransferHoverOver: "Add",
    DIALPADExternalTransferPopupHeader: "External Transfer",
  },
  serviceConfiguration: {
    outbound_call_flows: {
      default: {
        caller_id: "CALLER_ID",
      },
    },
  },
});
