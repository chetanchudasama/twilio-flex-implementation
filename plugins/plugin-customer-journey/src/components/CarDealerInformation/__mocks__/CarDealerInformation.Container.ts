import CarDealerInformation from "./CarDealerInformation";

export default CarDealerInformation;
