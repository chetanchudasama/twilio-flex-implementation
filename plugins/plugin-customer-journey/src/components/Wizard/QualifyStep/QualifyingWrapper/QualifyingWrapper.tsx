import React, { useEffect, useMemo, useState } from "react";

import { WizardStepWrapper, BaseWizardStepProps } from "@common/components";
import { Manager } from "@twilio/flex-ui";

import { preparePatchData } from "../../../../helpers/commonFunctions";
import { QualifyDetailModel } from "../../../../models/QualifyDetailModel";
import { useApplicationService } from "../../../../services/application.service";
import { useStaticService } from "../../../../services/static.service";
import { AppState } from "../../../../states";
import Loading from "../../../LoadingComponent/Loading";
import QualifyingContent from "../QualifyingContent/QualifyingContent.Container";
import { StateToProps, DispatchToProps } from "./QualifyingWrapper.Props";

export type ComponentProps = StateToProps &
  DispatchToProps &
  BaseWizardStepProps;

const QualifyingWrapper: React.FC<ComponentProps> = (props: ComponentProps) => {
  const { applicationId, nextStep } = props;
  const state: AppState = Manager.getInstance().store.getState();
  const token: string = useMemo(() => {
    return state.flex.session.ssoTokenPayload.token ?? "";
  }, [state.flex.session.ssoTokenPayload]);
  const phoenixToken: string = useMemo(() => {
    return state.crm?.crmState?.phoenixToken ?? "";
  }, [state.crm?.crmState?.phoenixToken]);

  const applicationService = useApplicationService(token);
  const staticService = useStaticService(token, phoenixToken);

  const [isLoading, setIsLoading] = useState<boolean>(false);
  const [qualifyingDetail, setQualifyingDetail] = useState<QualifyDetailModel>(
    new QualifyDetailModel()
  );
  const [errors, setErrors] = useState<Record<string, string[]>>({});
  const [isDataLoaded, setIsDataLoaded] = useState<boolean>();

  useEffect(() => {
    if (props.isStaticItemsFetched) {
      return;
    }

    staticService
      .getDropdownData()
      .then((response) => {
        props.setReasonForPurchaseItems(response.reasonForPurchaseItems);
        props.setTimeForPurchaseItems(response.timeForPurchaseItems);
        props.setWhereDidYouHearItems(response.whereDidYouHearItems);
        props.setStaticFetched(true);
      })
      .catch((error) => {
        console.error(error);
      });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const onNext = async () => {
    const errorObject: Record<string, string[]> = {};
    let hasErrors = false;

    console.warn("CHECKING NEXT BUTTON ERRORS");

    if (qualifyingDetail.drivingLicenseConfirmed === null) {
      errorObject.drivingLicenseConfirmed = [
        "Driving License is not confirmed",
      ];
      hasErrors = true;
    }
    if (qualifyingDetail.addressConfirmed === null) {
      errorObject.addressConfirmed = ["Address is not confirmed"];
      hasErrors = true;
    }
    if (qualifyingDetail.incomeConfirmed === null) {
      errorObject.incomeConfirmed = ["Income is not confirmed"];
      hasErrors = true;
    }
    if (qualifyingDetail.placeOfWorkConfirmed === null) {
      errorObject.placeOfWorkConfirmed = ["Place of work is not confirmed"];
      hasErrors = true;
    }
    if (qualifyingDetail.reasonForPurchase === null) {
      errorObject.reasonForPurchase = ["Reason for purchase is not selected"];
      hasErrors = true;
    }
    if (qualifyingDetail.timeForPurchase === null) {
      errorObject.timeForPurchase = ["Time for purchase is not selected"];
      hasErrors = true;
    }
    if (qualifyingDetail.whereDidYouHear === null) {
      errorObject.whereDidYouHear = ["This field is required"];
      hasErrors = true;
    }
    if (qualifyingDetail.hasCustomerAlreadyFoundCar === null) {
      errorObject.hasCustomerAlreadyFoundCar = [
        "Car already found is not confirmed",
      ];
      hasErrors = true;
    }

    // if (!qualifyingDetail.budgetInformation) {
    //   errorObject.maximumBorrowAmount = ["This field is required"];
    //   errorObject.monthlyBudgetAmount = ["This field is required"];
    //   hasErrors = true;
    // } else {
    //   const { budgetInformation } = qualifyingDetail;
    //   if (
    //     !budgetInformation.maximumBorrowAmount ||
    //     budgetInformation.maximumBorrowAmount <= 0
    //   ) {
    //     errorObject.maximumBorrowAmount = [
    //       "Amount should be greater than zero",
    //     ];
    //     hasErrors = true;
    //   }
    //   if (
    //     !budgetInformation.monthlyBudgetAmount ||
    //     budgetInformation.monthlyBudgetAmount <= 0
    //   ) {
    //     errorObject.monthlyBudgetAmount = [
    //       "Amount should be greater than zero",
    //     ];
    //     hasErrors = true;
    //   }
    // }
    setErrors(errorObject);

    console.warn("ERRORS", errorObject);

    if (hasErrors) {
      document.getElementById("please-confirm-header")?.scrollIntoView();
      return;
    }

    if (props.moveForward) {
      props.moveForward();
    }
  };

  const onPrevious = () => {
    if (props.moveBackward) {
      props.moveBackward();
    }
  };

  const fetchQualifyingDetails = () => {
    setIsLoading(true);
    setIsDataLoaded(false);
    applicationService
      .getQualifyingDetails(applicationId)
      .then((response) => {
        setQualifyingDetail(response);
      })
      .catch((error) => {
        console.error(error);
        setQualifyingDetail(new QualifyDetailModel());
      })
      .finally(() => {
        setIsLoading(false);
        setIsDataLoaded(true);
      });
  };

  useEffect(() => {
    if (applicationId > 0) {
      fetchQualifyingDetails();
    } else {
      setQualifyingDetail(new QualifyDetailModel());
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [applicationId]);

  const updateQualifyingDetails = (
    key: keyof QualifyDetailModel,
    value: string | boolean | any | null
  ) => {
    if (!applicationId) {
      return;
    }
    let condition = qualifyingDetail[key] !== value;

    if (typeof value === "object") {
      condition =
        JSON.stringify(qualifyingDetail[key]) !== JSON.stringify(value);
    }

    if (!condition) {
      return;
    }

    setIsLoading(true);

    console.log("UPDATE QUALIFYING DETAILS APP ID", applicationId);

    applicationService
      .updateQualifyingDetails(applicationId, [preparePatchData(key, value)])
      .then(() => {
        setQualifyingDetail({
          ...qualifyingDetail,
          [key]: value,
        });

        // reset validations for that key
        setErrors({
          ...errors,
          [key]: [],
        });
      })
      .catch((error) => {
        console.error(error);
      })
      .finally(() => {
        setIsLoading(false);
      });
  };

  return (
    <WizardStepWrapper
      onNext={onNext}
      onPrevious={onPrevious}
      disablePrevious
      nextStep={nextStep}
    >
      {isLoading && <Loading />}
      {isDataLoaded && (
        <QualifyingContent
          qualifyingDetail={qualifyingDetail}
          updateQualifyingDetails={updateQualifyingDetails}
          errors={errors}
        />
      )}
    </WizardStepWrapper>
  );
};

export default QualifyingWrapper;
