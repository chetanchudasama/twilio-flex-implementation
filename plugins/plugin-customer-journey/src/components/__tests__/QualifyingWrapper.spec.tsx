/* eslint-disable @typescript-eslint/no-non-null-assertion */
import { Manager } from "@twilio/flex-ui";
import { mount, ReactWrapper } from "enzyme";
import flushPromises from "flush-promises";
import React from "react";
import { act as Act } from "react-dom/test-utils";
import { create, act, ReactTestRenderer } from "react-test-renderer";
import QualifyingContent from "../Wizard/QualifyStep/QualifyingContent/QualifyingContent.Container";
import QualifyingWrapper, {
  ComponentProps,
} from "../Wizard/QualifyStep/QualifyingWrapper/QualifyingWrapper";
import { QualifyDetailModel } from "../../models/QualifyDetailModel";
import { ReasonForCarPurchaseItemModel } from "../../models/ReasonForCarPurchaseItemModel";
import { DropDownDataModel } from "../../models/DropDownDataModel";
import { TimeForPurchaseItemModel } from "../../models/TimeForPurchaseModel";
import { WhereDidYouHearItemModel } from "../../models/WhereDidYouHearItemModel";
import { flexManagerMock } from "../__mocks__/MockData";

jest.mock(
  "../Wizard/QualifyStep/QualifyingContent/QualifyingContent.Container"
);
jest.mock("../Wizard/QualifyStep/QualifyingContent/QualifyingContent");
jest.mock("../../services/application.service");
jest.mock("../../services/static.service");

const mockGetQualifyingDetails = jest.fn();
const mockUpdateQualifyingDetails = jest.fn();
const mockGetDropdownData = jest.fn();

jest.mock("../../services/application.service", () => ({
  useApplicationService: jest.fn(() => ({
    getQualifyingDetails: mockGetQualifyingDetails,
    updateQualifyingDetails: mockUpdateQualifyingDetails,
  })),
}));

jest.mock("../../services/static.service", () => ({
  useStaticService: jest.fn(() => ({
    getDropdownData: mockGetDropdownData,
  })),
}));

jest.mock("@common/components", () => {
  const components = jest.requireActual("@common/components");
  return {
    ...components,
    WizardStepWrapper: jest.fn(({ children, onNext, onPrevious }) => {
      return (
        <div>
          <button type="button" className="onNext" onClick={onNext}>
            onNext
          </button>
          <button type="button" className="onPrevious" onClick={onPrevious}>
            onPrevious
          </button>
          {children}
        </div>
      );
    }),
  };
});

let qualifyingDetails = new QualifyDetailModel();
const reasonForPurchaseItems: ReasonForCarPurchaseItemModel[] = [
  {
    id: 1,
    name: "reasonForPurchaseItem1",
  },
  {
    id: 2,
    name: "reasonForPurchaseItem2",
  },
  {
    id: 3,
    name: "reasonForPurchaseItem3",
  },
];
const timeForPurchaseItems: TimeForPurchaseItemModel[] = [
  {
    id: 1,
    name: "timeForPurchaseItem1",
  },
  {
    id: 2,
    name: "timeForPurchaseItem2",
  },
  {
    id: 3,
    name: "timeForPurchaseItem3",
  },
];
const whereDidYouHearItems: WhereDidYouHearItemModel[] = [
  {
    id: 1,
    name: "whereDidYouHearItem1",
  },
  {
    id: 2,
    name: "whereDidYouHearItem2",
  },
  {
    id: 3,
    name: "whereDidYouHearItem3",
  },
];
const dropDownDataItems = new DropDownDataModel();

Manager.getInstance = flexManagerMock;

const defaultProps = {
  isStaticItemsFetched: true,
  applicationId: 1,
  setReasonForPurchaseItems: jest.fn(),
  setStaticFetched: jest.fn(),
  setTimeForPurchaseItems: jest.fn(),
  setWhereDidYouHearItems: jest.fn(),
};

let props: ComponentProps = defaultProps;

describe("QualifyingWrapper", () => {
  beforeEach(() => {
    props = defaultProps;
    dropDownDataItems.reasonForPurchaseItems = reasonForPurchaseItems;
    dropDownDataItems.timeForPurchaseItems = timeForPurchaseItems;
    dropDownDataItems.whereDidYouHearItems = whereDidYouHearItems;

    qualifyingDetails = Object.assign(new QualifyDetailModel(), {
      drivingLicenseConfirmed: true,
      addressConfirmed: true,
      incomeConfirmed: true,
      placeOfWorkConfirmed: true,
      reasonForPurchase: reasonForPurchaseItems[0],
      timeForPurchase: timeForPurchaseItems[0],
      whereDidYouHear: whereDidYouHearItems[0],
      hasCustomerAlreadyFoundCar: true,
    });
    mockGetQualifyingDetails.mockImplementationOnce(() =>
      Promise.resolve(qualifyingDetails)
    );
    mockGetDropdownData.mockImplementationOnce(() =>
      Promise.resolve(dropDownDataItems)
    );
  });

  afterEach(() => {
    mockGetQualifyingDetails.mockClear();
    mockUpdateQualifyingDetails.mockClear();
    mockGetDropdownData.mockClear();

    jest.clearAllMocks();
    jest.clearAllTimers();
  });

  it("renders correctly, without any props", async () => {
    let wrapper: ReactTestRenderer;
    await act(async () => {
      wrapper = create(<QualifyingWrapper {...props} />);
      await flushPromises();
    });
    expect(wrapper!.toJSON()).toMatchSnapshot();
  });

  it("renders correctly, with props", async () => {
    props.moveForward = jest.fn();
    props.moveBackward = jest.fn();
    let wrapper: ReactTestRenderer;
    await act(async () => {
      wrapper = create(<QualifyingWrapper {...props} />);
      await flushPromises();
    });
    expect(mockGetQualifyingDetails).toHaveBeenCalled();
    expect(wrapper!.toJSON()).toMatchSnapshot();
  });
  it("if qualifying details is valid, injected moveForward should be called", async () => {
    const moveForward = jest.fn();
    props.moveForward = moveForward;

    let wrapper: ReactWrapper<any, Readonly<any>, React.Component>;
    await Act(async () => {
      wrapper = mount(<QualifyingWrapper {...props} />);
      await flushPromises();
    });
    wrapper!.update();
    expect(mockGetQualifyingDetails).toHaveBeenCalled();

    wrapper!.find(".onNext").last().simulate("click");

    expect(moveForward).toHaveBeenCalled();
  });

  it("when previous button clicked and move backward event is provided, move backward event should be emitted", async () => {
    const moveBackward = jest.fn();
    props.moveBackward = moveBackward;
    await Act(async () => {
      const wrapper = mount(<QualifyingWrapper {...props} />);
      await flushPromises();
      wrapper.find(".onPrevious").last().simulate("click");

      expect(moveBackward).toHaveBeenCalledTimes(1);
    });
  });

  it("qualifying details should not be fetched, if applicationId less than zero", async () => {
    const updatedProps = { ...props };
    updatedProps.applicationId = -1;

    await Act(async () => {
      const wrapper = mount(<QualifyingWrapper {...updatedProps} />);
      await flushPromises();
      wrapper.update();

      expect(mockGetQualifyingDetails).not.toHaveBeenCalled();
    });
  });

  it("qualifying details should be fetched, when application id is changed", async () => {
    await Act(async () => {
      const wrapper = mount(<QualifyingWrapper {...props} />);
      await flushPromises();
      wrapper.update();
    });
    expect(mockGetQualifyingDetails).toHaveBeenCalled();
    expect(mockGetQualifyingDetails).toHaveBeenCalledWith(1);
  });

  it("should fetch dropdownData if not fetched", async () => {
    const moveBackward = jest.fn();
    props.moveBackward = moveBackward;
    props.isStaticItemsFetched = false;

    mockUpdateQualifyingDetails.mockImplementationOnce(() =>
      Promise.resolve(null)
    );

    await Act(async () => {
      const wrapper = mount(<QualifyingWrapper {...props} />);
      wrapper.update();
      await flushPromises();
      wrapper.update();
    });

    expect(mockGetDropdownData).toHaveBeenCalledTimes(1);
  });

  it("should not fetch dropdownData if data already stored in store", async () => {
    const moveBackward = jest.fn();
    props.moveBackward = moveBackward;
    props.isStaticItemsFetched = true;

    mockUpdateQualifyingDetails.mockImplementationOnce(() =>
      Promise.resolve(null)
    );

    await Act(async () => {
      const wrapper = mount(<QualifyingWrapper {...props} />);
      wrapper.update();
      await flushPromises();
      wrapper.update();
    });

    expect(mockGetDropdownData).not.toHaveBeenCalled();
  });

  it("while updating qualifying details; if previous data is same as current data, then it should not update the data", async () => {
    const moveBackward = jest.fn();
    props.moveBackward = moveBackward;

    let wrapper: ReactWrapper<any, Readonly<any>, React.Component>;

    mockUpdateQualifyingDetails.mockImplementationOnce(() => Promise.resolve());

    await Act(async () => {
      wrapper = mount(<QualifyingWrapper {...props} />);
      await flushPromises();
    });
    wrapper!.update();

    const QualifyingContentComponent = wrapper!.find(QualifyingContent).last();

    expect(QualifyingContentComponent.exists()).toBeTruthy();

    QualifyingContentComponent.invoke("updateQualifyingDetails")(
      "drivingLicenseConfirmed",
      qualifyingDetails.drivingLicenseConfirmed
    );
    wrapper!.update();
    await flushPromises();
    wrapper!.update();

    expect(mockUpdateQualifyingDetails).not.toHaveBeenCalled();
  });

  it("while updating qualifying details; if previous data is not same as current data, then it should update the data", async () => {
    const moveBackward = jest.fn();
    props.moveBackward = moveBackward;

    mockUpdateQualifyingDetails.mockImplementationOnce(() => Promise.resolve());

    let wrapper: ReactWrapper<any, Readonly<any>, React.Component>;

    await Act(async () => {
      wrapper = mount(<QualifyingWrapper {...props} />);
      await flushPromises();
    });
    wrapper!.update();

    const QualifyingContentComponent = wrapper!.find(QualifyingContent).last();

    expect(QualifyingContentComponent.exists()).toBeTruthy();

    await Act(async () => {
      QualifyingContentComponent.invoke("updateQualifyingDetails")(
        "drivingLicenseConfirmed",
        false
      );
      wrapper!.update();
      await flushPromises();
    });
    wrapper!.update();

    expect(mockUpdateQualifyingDetails).toHaveBeenCalledTimes(1);
  });
});
