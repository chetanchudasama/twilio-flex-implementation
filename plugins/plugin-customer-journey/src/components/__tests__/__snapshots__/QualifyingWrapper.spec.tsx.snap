// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QualifyingWrapper renders correctly, with props 1`] = `
<div>
  <button
    className="onNext"
    onClick={[Function]}
    type="button"
  >
    onNext
  </button>
  <button
    className="onPrevious"
    onClick={[Function]}
    type="button"
  >
    onPrevious
  </button>
  <div />
</div>
`;

exports[`QualifyingWrapper renders correctly, without any props 1`] = `
<div>
  <button
    className="onNext"
    onClick={[Function]}
    type="button"
  >
    onNext
  </button>
  <button
    className="onPrevious"
    onClick={[Function]}
    type="button"
  >
    onPrevious
  </button>
  <div />
</div>
`;
